File: Problem/By section/Greedy/Problem 406/problem_406.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    static bool sortcol(const vector<int>& a, const vector<int>& b) {
        if(a[0] == b[0]) return a[1] < b[1];
        return a[0] > b[0];
    }
    
    // Time: O(n^2), Space: O(n)
    vector<vector<int>> reconstructQueue(vector<vector<int>>& people) {
        if(people.size() == 0) return vector<vector<int>>{};

        sort(people.begin(), people.end(), sortcol);

        vector<vector<int>> result(people.size());
        int cnt = 0;
        
        for(auto person : people) {
            for(int i = cnt; i > person[1]; i--) {
                result[i] = result[i - 1];
            }
            
            cnt++;
            result[person[1]] = person;
        }
        

        return result;
    }
};

int main(int argc, char* argv[]) {
    return 0;
}

File: Problem/By section/Greedy/Problem 406/problem_406.java
import java.util.ArrayList;
import java.util.Arrays;

class Solution {
    // Time: O(nlogn), Space: O(n)
    public int[][] reconstructQueue(int[][] people) {
        if(people.length == 0) return new int[0][0];

        // sort the people array by its height in descending order
        // if meet people with same height, sort them by their second index in increasing order
        Arrays.sort(people, (arr1, arr2) -> ((arr1[0] == arr2[0]) ? arr1[1] - arr2[1] : -(arr1[0] - arr2[0])));

        // start with the highest people
        // insert them by the second index
        // since the height is from high to low
        // we will always construct a valid subarray with
        // the element add as the shorest people
        ArrayList<int[]> ans = new ArrayList<>();
        for(int[] i : people){
            ans.add(i[1], i);
        }

        return ans.toArray(new int[people.length][]);
    }
}